database/engine2/metric.c

```diff
@@ -0,0 +1,636 @@
+    if(!metric->first_time_t) {
+        metric->first_time_t = first_time_t;
+
+//        if(unlikely(metric->latest_time_t_clean < metric->first_time_t))
+//            metric->latest_time_t_clean = metric->first_time_t;
+//
+//        if(unlikely(metric->latest_time_t_hot < metric->first_time_t))
+//            metric->latest_time_t_hot = metric->first_time_t;
+
+        ret = true;
+    }
+    netdata_spinlock_unlock(&metric->timestamps_lock);
+
+    return ret;
+}
+
+time_t mrg_metric_get_first_time_t(MRG *mrg __maybe_unused, METRIC *metric) {
+    time_t first_time_t;
+    netdata_spinlock_lock(&metric->timestamps_lock);
+    first_time_t = metric->first_time_t;
+    if(!first_time_t) {
+        if(metric->latest_time_t_clean)
+            first_time_t = metric->latest_time_t_clean;
+
+        if(!first_time_t || metric->latest_time_t_hot < metric->latest_time_t_clean)
+            first_time_t = metric->latest_time_t_hot;
+    }
+    netdata_spinlock_unlock(&metric->timestamps_lock);
+
+    return first_time_t;
+}
+
+bool mrg_metric_set_clean_latest_time_t(MRG *mrg __maybe_unused, METRIC *metric, time_t latest_time_t) {
+    netdata_spinlock_lock(&metric->timestamps_lock);
+
+//    internal_fatal(latest_time_t > now_realtime_sec(),
+//                   "DBENGINE METRIC: metric latest time is in the future");
+
+    internal_fatal(metric->latest_time_t_clean > latest_time_t,
+                   "DBENGINE METRIC: metric new clean latest time is older than the previous one");
+
+    metric->latest_time_t_clean = latest_time_t;
+
+    if(unlikely(!metric->first_time_t))
+        metric->first_time_t = latest_time_t;
+
+//    if(unlikely(metric->first_time_t > latest_time_t))
+//        metric->first_time_t = latest_time_t;
+
+    netdata_spinlock_unlock(&metric->timestamps_lock);
+    return true;
+}
+
+bool mrg_metric_set_hot_latest_time_t(MRG *mrg __maybe_unused, METRIC *metric, time_t latest_time_t) {
+//    internal_fatal(latest_time_t > now_realtime_sec(),
+//                   "DBENGINE METRIC: metric latest time is in the future");
+
+    netdata_spinlock_lock(&metric->timestamps_lock);
+    metric->latest_time_t_hot = latest_time_t;
+
+    if(unlikely(!metric->first_time_t))
+        metric->first_time_t = latest_time_t;
+
+//    if(unlikely(metric->first_time_t > latest_time_t))
+//        metric->first_time_t = latest_time_t;
+
+    netdata_spinlock_unlock(&metric->timestamps_lock);
+    return true;
+}
+
+time_t mrg_metric_get_latest_time_t(MRG *mrg __maybe_unused, METRIC *metric) {
+    time_t max;
+    netdata_spinlock_lock(&metric->timestamps_lock);
+    max = MAX(metric->latest_time_t_clean, metric->latest_time_t_hot);
+    netdata_spinlock_unlock(&metric->timestamps_lock);
+    return max;
+}
+
+bool mrg_metric_set_update_every(MRG *mrg __maybe_unused, METRIC *metric, time_t update_every) {
+    if(!update_every)
+        return false;
+
+    netdata_spinlock_lock(&metric->timestamps_lock);
```

Comment:
```sh
/database/engine2/metric.c: 375 in mrg_metric_set_update_every()
369     
370     bool mrg_metric_set_update_every(MRG *mrg __maybe_unused, METRIC *metric, time_t update_every) {
371         if(!update_every)
372             return false;
373     
374         netdata_spinlock_lock(&metric->timestamps_lock);
>>>     CID 433488:  High impact quality  (Y2K38_SAFETY)
>>>     A "time_t" value is stored in an integer with too few bits to accommodate it.  The expression "update_every" is cast to "uint32_t".
375         metric->latest_update_every = update_every;
376         netdata_spinlock_unlock(&metric->timestamps_lock);
377     
378         return true;
379     }
380    
```

Reply:
Ignore it. update every cannot get that big.